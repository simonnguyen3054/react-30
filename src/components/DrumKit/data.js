export const data = [
  {
    key: 65,
    letter: "A",
    type: "clap"
  },
  {
    key: 83,
    letter: "S",
    type: "hihat"
  },
  {
    key: 68,
    letter: "D",
    type: "kick"
  },
  {
    key: 70,
    letter: "F",
    type: "openhat"
  },
  {
    key: 71,
    letter: "G",
    type: "boom"
  },
  {
    key: 72,
    letter: "H",
    type: "ride"
  },
  {
    key: 74,
    letter: "J",
    type: "snare"
  },
  {
    key: 75,
    letter: "K",
    type: "tom"
  },
  {
    key: 76,
    letter: "L",
    type: "tink"
  }
];
