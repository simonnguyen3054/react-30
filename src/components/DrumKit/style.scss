.drumkit_background {
  background: linear-gradient(#c7c7c778, #726f6fab, transparent),
    url("../../assets/images/drum_kit_bg.jpg") no-repeat center;
  background-size: cover;
  min-height: 100vh;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;

  //hide it for now
  display: none;

  .keys {
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(12rem, auto));
    //auto-fit wrap columns into rows when thereâ€™s not enough space in our viewport to fit the 12rem minimum value without overflowing the content.
    //auto for max value is to set the column width based on the width set on its children
    row-gap: 3rem;
    justify-content: center;
  }

  .key {
    border: 0.2rem solid black;
    font-size: 1.5rem;
    padding: 1rem 0.5rem;
    transition: all 0.07s ease;
    width: 10rem;
    text-align: center;
    color: white;
    background: rgba(0, 0, 0, 0.4);
    text-shadow: 0 0 0.5rem black;
    justify-self: center;
  }

  kbd {
    display: block;
    font-size: 4rem;
  }

  .playing {
    transform: scale(1.1);
    border-color: #fcb92f;
    box-shadow: 0 0 1rem #fcb92f;
  }

  .sound {
    font-size: 1.2rem;
    text-transform: uppercase;
    letter-spacing: 0.1rem;
    color: #fcb92f;
  }
}
