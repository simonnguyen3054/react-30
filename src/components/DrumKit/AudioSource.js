import React, { useEffect, useRef } from "react";
import "./style.scss";

const AudioSource = ({
  src,
  encoding = "audio/mpeg",
  keyCode,
  handleKeyOnPlayed
}) => {
  const ref = useRef();
  useEffect(() => {
    window.addEventListener("keydown", onKeyPress);
  }, []);

  const onKeyPress = event => {
    if (keyCode === event.keyCode) {
      const audio = ref.current;
      handleKeyOnPlayed(event.keyCode); //return the keyCode that's playing
      audio.currentTime = 0; //rewind to the start
      return playAudio(audio);
    }
  };

  //handle Promise resulting from audio.play()
  async function playAudio(audio) {
    try {
      await audio.play();
    } catch (err) {
      console.log(err);
    }
  }

  return <audio src={src} ref={ref} type={encoding}></audio>;
};

export default AudioSource;
