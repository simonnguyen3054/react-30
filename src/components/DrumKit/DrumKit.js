import React, { useEffect, useState, useRef, createRef } from "react";
import { data as Data } from "./data";
import AudioSource from "./AudioSource";
import "./style.scss";
import clsx from "clsx";
import {
  clap,
  boom,
  hihat,
  kick,
  openhat,
  ride,
  snare,
  tink,
  tom
} from "../../assets/sounds";

const DrumKit = () => {
  const audioHashMap = {
    clap,
    boom,
    hihat,
    kick,
    openhat,
    ride,
    snare,
    tink,
    tom
  };
  const keysRef = useRef([]); //create an empty array to store div.keys elements
  const [pressedKeyCode, setPressedKeyCode] = useState();
  useEffect(() => {
    //check when transitionend, remove the classname "playing" from div.key element
    keysRef.current.forEach(key =>
      key.addEventListener("transitionend", removeTransition)
    );
  }, []);
  const removeTransition = event => {
    if (event.propertyName !== "transform") return; //skip if not a transform
    setPressedKeyCode(null);
  };

  return (
    <div className="drumkit_background">
      <div className="keys">
        {Data.map((item, i) => (
          <div
            key={item.key}
            className={clsx("key", pressedKeyCode === item.key && "playing")} //add classname "playing" on a pressed div.key element
            ref={key => (keysRef.current[i] = key)} //assign a list of div.key elements
          >
            <kbd>{item.letter}</kbd>
            <span className="sound">{item.type}</span>
            <AudioSource
              keyCode={item.key}
              src={audioHashMap[item.type]}
              handleKeyOnPlayed={keyCode => setPressedKeyCode(keyCode)}
            />
          </div>
        ))}
      </div>
    </div>
  );
};

export default DrumKit;
