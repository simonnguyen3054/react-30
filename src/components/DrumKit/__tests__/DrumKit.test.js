import React from "react";
import { mount } from "enzyme";
import DrumKit from "../DrumKit";
import AudioSource from "../AudioSource";

let wrapper = null;
const setPressedKeyCode = jest.fn();
const useStateSpy = jest.spyOn(React, "useState");
useStateSpy.mockImplementation(pressedKeyCode => [
  pressedKeyCode,
  setPressedKeyCode
]);

beforeEach(() => {
  wrapper = mount(<DrumKit />);
});

afterEach(() => {
  wrapper.unmount();
});

describe("Drumkit", () => {
  it("should render all drum kits", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
