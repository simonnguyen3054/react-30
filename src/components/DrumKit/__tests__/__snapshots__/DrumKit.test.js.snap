// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drumkit should render all drum kits 1`] = `
<DrumKit>
  <div
    className="drumkit_background"
  >
    <div
      className="keys"
    >
      <div
        className="key"
        key="65"
      >
        <kbd>
          A
        </kbd>
        <span
          className="sound"
        >
          clap
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={65}
          src="clap.wav"
        >
          <audio
            src="clap.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="83"
      >
        <kbd>
          S
        </kbd>
        <span
          className="sound"
        >
          hihat
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={83}
          src="hihat.wav"
        >
          <audio
            src="hihat.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="68"
      >
        <kbd>
          D
        </kbd>
        <span
          className="sound"
        >
          kick
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={68}
          src="kick.wav"
        >
          <audio
            src="kick.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="70"
      >
        <kbd>
          F
        </kbd>
        <span
          className="sound"
        >
          openhat
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={70}
          src="openhat.wav"
        >
          <audio
            src="openhat.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="71"
      >
        <kbd>
          G
        </kbd>
        <span
          className="sound"
        >
          boom
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={71}
          src="boom.wav"
        >
          <audio
            src="boom.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="72"
      >
        <kbd>
          H
        </kbd>
        <span
          className="sound"
        >
          ride
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={72}
          src="ride.wav"
        >
          <audio
            src="ride.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="74"
      >
        <kbd>
          J
        </kbd>
        <span
          className="sound"
        >
          snare
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={74}
          src="snare.wav"
        >
          <audio
            src="snare.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="75"
      >
        <kbd>
          K
        </kbd>
        <span
          className="sound"
        >
          tom
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={75}
          src="tom.wav"
        >
          <audio
            src="tom.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
      <div
        className="key"
        key="76"
      >
        <kbd>
          L
        </kbd>
        <span
          className="sound"
        >
          tink
        </span>
        <AudioSource
          handleKeyOnPlayed={[Function]}
          keyCode={76}
          src="tink.wav"
        >
          <audio
            src="tink.wav"
            type="audio/mpeg"
          />
        </AudioSource>
      </div>
    </div>
  </div>
</DrumKit>
`;
